crates/delta/src/routes/channels/voice_join.rs

```diff
@@ -1,105 +1,66 @@
 use revolt_config::config;
-use revolt_database::{
-    util::{permissions::DatabasePermissionQuery, reference::Reference},
-    Channel, Database, User,
-};
 use revolt_models::v0;
+use revolt_database::{util::{permissions::perms, reference::Reference}, voice::{get_channel_node, raise_if_in_voice, VoiceClient}, Channel, Database, SystemMessage, User, AMQP};
 use revolt_permissions::{calculate_channel_permissions, ChannelPermission};
 use revolt_result::{create_error, Result};
+
 use rocket::{serde::json::Json, State};
 
 /// # Join Call
 ///
 /// Asks the voice server for a token to join the call.
 #[openapi(tag = "Voice")]
-#[post("/<target>/join_call")]
-pub async fn call(
-    db: &State<Database>,
-    user: User,
-    target: Reference,
-) -> Result<Json<v0::LegacyCreateVoiceUserResponse>> {
-    let channel = target.as_channel(db).await?;
-    let mut query = DatabasePermissionQuery::new(db, &user).channel(&channel);
-    calculate_channel_permissions(&mut query)
-        .await
-        .throw_if_lacking_channel_permission(ChannelPermission::Connect)?;
+#[post("/<target>/join_call", data="<data>")]
+pub async fn call(db: &State<Database>, amqp: &State<AMQP>, voice: &State<VoiceClient>, user: User, target: Reference, data: Json<v0::DataJoinCall>) -> Result<Json<v0::CreateVoiceUserResponse>> {
+    if !voice.is_enabled() {
+        return Err(create_error!(LiveKitUnavailable))
+    }
 
     let config = config().await;
-    if config.api.security.voso_legacy_token.is_empty() {
-        return Err(create_error!(VosoUnavailable));
-    }
 
-    match channel {
-        Channel::SavedMessages { .. } | Channel::TextChannel { .. } => {
-            return Err(create_error!(CannotJoinCall))
-        }
-        _ => {}
-    }
+    let channel = target.as_channel(db).await?;
+
+    raise_if_in_voice(&user, channel.id()).await?;
```

Comment:
Instead of raising, we could end the previous session and start the new one? Think "I joined the call on my phone and now i'm hopping on my computer". Maybe that can be a param flag?